Unstable support for MSC 3489 live location sharing

torrybr opened the pull request:
This merge request introduces the live location sharing feature as per [MSC3489](https://github.com/matrix-org/matrix-spec-proposals/pull/3489).

These changes allow users to share their live location, start and stop the sharing, and check if the location sharing is live.

jplatte commented on crates/ruma-events/src/live_location.rs:
```diff
@@ -0,0 +1,113 @@
+/// event. It contains information about a live location sharing event.
+#[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[ruma_event(type = "org.matrix.msc3672.beacon_info", alias = "m.beacon", kind = State, state_key_type = OwnedUserId)]
```
This seems like a typo, right?

```suggestion
#[ruma_event(type = "org.matrix.msc3672.beacon_info", alias = "m.beacon_info", kind = State, state_key_type = OwnedUserId)]
```

torrybr replied:
Thanks for the feedback! I will make those changes and get it committed.

jplatte reviewed (commented):
Hi, and thanks for the PR!

Please note that all of the modules in `ruma-events` that define event (content) types are named after the event type. That means that the `live_location` module you added should be split into two modules `beacon` and `beacon_info`.

jplatte commented on crates/ruma-events/src/beacon.rs:
```diff
@@ -0,0 +1,77 @@
+    /// `timeout` is a u64 that represents the length of time in milliseconds that the location
+    /// will be live. So the location will stop being shared at `m.ts + timeout` milliseconds
+    /// since the epoch.
+    pub timeout: MilliSecondsSinceUnixEpoch,
```
This should be either a `std::time::Duration` (using `#[serde(with = "ruma_common::serde::duration::ms")]`), or `js_int::UInt` since it's not a timestamp which is what `MilliSecondsSinceUnixEpoch` represents.

jplatte commented on crates/ruma-events/src/beacon.rs:
```diff
@@ -0,0 +1,77 @@
+            return false;
+        }
+
+        return self.ts.unwrap().get() + self.timeout.get() < now_ts.get();
```
No need for a `return` at the end of the function, and also the `self.live` check can be part of the expression instead of a separate early return.

torrybr replied:
Good catch, first Rust PR of my life so learning a lot, appreciate you reviewing!

mars1999 committed:
feat(beacon): add unstable support for msc-3489

Add missing even type to beacon handling

This is a work in progress.

move beacon struct to `beacon` module

wip

cleanup

make it compile; add test

ensure that location is enabled for beacon live sharing

use beacon_info instead of beacon for state event

cleanup

feat(beacon): add unstable support for msc-3489

feat(beacon): add unstable support for msc-3489

squash commits

refactor(beacon_info): rename files to be consistent with spec name

tests(beacon_info): Update beacon_info test for serialization and de_serialization

revert(pinnedevents): remove accdiental change

chore: update formatting

docs(beacon_info): update descriptions

clean up tests

refactor(beacon_info): update file naming and paths

feat(beacon): integrate beacon event with tests

test(beacon_info): fix broken test for creating asset type

refactor(beacon_info): update constructor params for owned_event_id and geo_uri

fix(enums): beacon info import fix

remove unused imports

fix(beacon_info): bad naming alias

fix(beacon_info): set the default ts time on creation

fix(beacon_info): fix tests for calculating expiration

remove some docs

docs(changelog): update changelog

fix clippy

use passed in value or now

fix clippy

fix tests

update docs

update correct changelog

revert changelog

revert changes

update changelog

torrybr commented:
@jplatte This has been revamped and should be ready for a second look if you have some spare time!

torrybr committed:
test(beacon_info): add test for start and stop functions

torrybr committed:
format(beacon_info): fix formatting

torrybr committed:
test(beacon_info): update tests to use assert

zecakeh commented on crates/ruma-events/CHANGELOG.md:
```diff
@@ -3,6 +3,9 @@
 
  - Add support for encrypted stickers as sent by several bridges under the flag `compat-encrypted-stickers`
+ - Add unstable support
+   for [MSC3489](https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/location-streaming/proposals/3489-location-streaming.md)
```
We don’t usually link to the MSC in the changelog, but in the docs of the module/event. Also please link to the PR rather than the rendered file.

zecakeh commented on crates/ruma-events/Cargo.toml:
```diff
@@ -33,6 +33,7 @@ unstable-msc3291 = []
 unstable-msc3381 = ["unstable-msc1767"]
 unstable-msc3401 = []
 unstable-msc3488 = ["unstable-msc1767"]
+unstable-msc3489 = ["unstable-msc1767", "unstable-msc3488"]
```
We don’t need to enable `unstable-msc1767` here, it’s already done by MSC3488.

zecakeh commented on crates/ruma-events/src/beacon.rs:
```diff
@@ -0,0 +1,41 @@
+use ruma_macros::EventContent;
+use serde::{Deserialize, Serialize};
+
+/// `BeaconEventContent` is a struct that represents the content of a beacon
```
Too verbose:

```suggestion
/// The content of a beacon
```

zecakeh commented on crates/ruma-events/src/beacon.rs:
```diff
@@ -0,0 +1,41 @@
+}
+
+impl BeaconEventContent {
+    /// Create a new `BeaconEventContent` with the given beacon_info_event_id and geo_uri
```
We should document what happens if the ts is not provided.

zecakeh commented on crates/ruma-events/src/enums.rs:
```diff
@@ -88,6 +88,9 @@ event_enum! {
         #[cfg(feature = "unstable-msc3381")]
         #[ruma_enum(ident = UnstablePollEnd)]
         "org.matrix.msc3381.poll.end" => super::poll::unstable_end,
+        #[cfg(all(feature = "unstable-msc3489", feature = "unstable-msc3488"))]
```
Since `unstable-msc3489` enables `unstable-msc3488`, we only need to use the former.

zecakeh commented on crates/ruma-events/src/enums.rs:
```diff
@@ -127,6 +130,9 @@ event_enum! {
         "m.room.topic" => super::room::topic,
         "m.space.child" => super::space::child,
         "m.space.parent" => super::space::parent,
+        #[cfg(all(feature = "unstable-msc3489", feature = "unstable-msc3488"))]
```
Same here.

zecakeh commented on crates/ruma-events/src/enums.rs:
```diff
@@ -307,6 +313,8 @@ impl AnyMessageLikeEventContent {
     /// This is a helper function intended for encryption. There should not be a reason to access
     /// `m.relates_to` without first destructuring an `AnyMessageLikeEventContent` otherwise.
     pub fn relation(&self) -> Option<encrypted::Relation> {
+        #[cfg(all(feature = "unstable-msc3489", feature = "unstable-msc3488"))]
```
Same here.

zecakeh commented on crates/ruma-events/src/enums.rs:
```diff
@@ -359,6 +367,10 @@ impl AnyMessageLikeEventContent {
             | Self::UnstablePollEnd(UnstablePollEndEventContent { relates_to, .. }) => {
                 Some(encrypted::Relation::Reference(relates_to.clone()))
             }
+            #[cfg(all(feature = "unstable-msc3489", feature = "unstable-msc3488"))]
```
Same here.

zecakeh commented on crates/ruma-events/src/room.rs:
```diff
@@ -13,6 +13,10 @@ use serde::{de, Deserialize, Serialize};
 pub mod avatar;
+
+#[cfg(feature = "unstable-msc3489")]
+pub mod beacon_info;
```
It should not be here but at the top level. It would be here if it was named `m.room.beacon_info`.

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+
+use crate::location::AssetContent;
+
+/// `BeaconInfoEventContent` is a struct that represents the content of a beacon_info state
```
Too verbose:

```suggestion
/// The content of a beacon_info state
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// `live` is a boolean that should be true when a user starts sharing location.
```
```suggestion
    /// Whether the user starts sharing their location.
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+    /// `live` is a boolean that should be true when a user starts sharing location.
+    pub live: bool,
+
+    /// The time when location sharing started, in milliseconds since the epoch.
```
We don’t need to specify the unit, it’s clear from the type.

```suggestion
    /// The time when location sharing started.
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+
+    /// `timeout` represents the length of time in milliseconds that the location
+    /// will be live. So the location will stop being shared at `m.ts + timeout` milliseconds
+    /// since the epoch.
```
```suggestion
    /// The duration that the location sharing will be live.
    ///
    /// Meaning that the location will stop being shared at `ts + timeout`.
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+    #[serde(default, with = "ruma_common::serde::duration::ms")]
+    pub timeout: Duration,
+
+    /// `asset` is an [`AssetContent`] that this message refers to.
```
```suggestion
    /// The asset that this message refers to.
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    /// starts the beacon being live.
```
```suggestion
    /// Starts the beacon being live.
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+    ///
+    /// This method calculates the current time and compares it with the beacon's start time plus
+    /// its timeout. If the beacon is not live or the current time is greater than the beacon's
+    /// start time plus its timeout, it returns false, indicating that the beacon is not live.
```
```suggestion
    /// start time plus its timeout, it returns `false`, indicating that the beacon is not live.
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+    /// This method calculates the current time and compares it with the beacon's start time plus
+    /// its timeout. If the beacon is not live or the current time is greater than the beacon's
+    /// start time plus its timeout, it returns false, indicating that the beacon is not live.
+    /// Otherwise, it returns true.
```
```suggestion
    /// Otherwise, it returns `true`.
```

zecakeh commented on crates/ruma-events/src/room/beacon_info.rs:
```diff
@@ -0,0 +1,81 @@
+    pub fn is_live(&self) -> bool {
+        self.live
+            && self.ts.get() + UInt::try_from(self.timeout.as_millis()).unwrap()
+                > MilliSecondsSinceUnixEpoch::now().get()
```
This seems cleaner:

```suggestion
            && self.ts.to_system_time().and_then(|t| t.checked_add(self.timeout)).is_some_and(|t| t > SystemTime::now())
```

zecakeh commented on crates/ruma-events/tests/it/beacon.rs:
```diff
@@ -0,0 +1,48 @@
+#![cfg(all(feature = "unstable-msc3489", feature = "unstable-msc3488"))]
```
Same as previously.

zecakeh commented on crates/ruma-events/tests/it/beacon_info.rs:
```diff
@@ -0,0 +1,127 @@
+#![cfg(all(feature = "unstable-msc3489", feature = "unstable-msc3488"))]
```
Same as previously.

zecakeh commented on crates/ruma/Cargo.toml:
```diff
@@ -205,6 +205,7 @@ unstable-msc3246 = ["ruma-events?/unstable-msc3246"]
 unstable-msc3266 = ["ruma-client-api?/unstable-msc3266"]
 unstable-msc3291 = ["ruma-events?/unstable-msc3291"]
 unstable-msc3381 = ["ruma-events?/unstable-msc3381"]
+unstable-msc3489 = ["ruma-events?/unstable-msc3489"]
```
It should be below `unstable-msc3488`.

zecakeh commented on crates/ruma/Cargo.toml:
```diff
@@ -259,6 +260,7 @@ __ci = [
     "unstable-msc3266",
     "unstable-msc3291",
     "unstable-msc3381",
+    "unstable-msc3489",
```
Same.

torrybr committed:
docs(beacon): update docs, move files to reflect MR feedback

zecakeh commented on crates/ruma-events/CHANGELOG.md:
```diff
@@ -3,6 +3,8 @@
 Improvements:
 
  - Add support for encrypted stickers as sent by several bridges under the flag `compat-encrypted-stickers`
+ - Add unstable support for `m.beacon` & `m.beacon_info` events
```
Sorry, I meant you can still mention the MSC, just avoid to link to it.

zecakeh commented on crates/ruma-events/src/beacon.rs:
```diff
@@ -0,0 +1,43 @@
+//! Types for the `org.matrix.msc3489.beacon` event, the unstable version of
+//! `m.beacon`.
+//!
+//! [MSC3489]: https://github.com/matrix-org/matrix-spec-proposals/pull/3489
```
This is a link reference. The link should be included somewhere in the text to appear. For example:

```rust
//! Types for the `org.matrix.msc3489.beacon` event, the unstable version of
//! `m.beacon` ([MSC3489]).
```

zecakeh commented on crates/ruma-events/src/beacon_info.rs:
```diff
@@ -0,0 +1,82 @@
+//! Types for the `org.matrix.msc3489.beacon_info` state event, the unstable version of
+//! `m.beacon_info`.
+//!
+//! [MSC3489]: https://github.com/matrix-org/matrix-spec-proposals/pull/3489
```
Same here

zecakeh commented on crates/ruma-events/src/beacon.rs:
```diff
@@ -0,0 +1,43 @@
+use ruma_macros::EventContent;
+use serde::{Deserialize, Serialize};
+
+/// The content of a beacon
```
```suggestion
/// The content of a beacon.
```

zecakeh commented on crates/ruma-events/src/beacon_info.rs:
```diff
@@ -0,0 +1,82 @@
+
+use crate::location::AssetContent;
+
+/// The content of a beacon_info state
```
```suggestion
/// The content of a beacon_info state.
```

zecakeh commented on crates/ruma-events/src/beacon_info.rs:
```diff
@@ -0,0 +1,82 @@
+}
+
+impl BeaconInfoEventContent {
+    /// Creates a new `BeaconInfoEventContent` with the given description, live, timeout and asset.
```
Same thing as above, the behavior of not providing the ts should be documented.

zecakeh commented on crates/ruma-events/tests/it/beacon_info.rs:
```diff
@@ -0,0 +1,127 @@
+}
+
+#[test]
+fn beacon_info_start_event_content_serialization() {
```
Could we have tests of full event (`BeaconInfoEvent`) serialization and deserialization too?

zecakeh commented on crates/ruma-events/tests/it/beacon.rs:
```diff
@@ -0,0 +1,48 @@
+}
+
+#[test]
+fn beacon_event_content_serialization() {
```
Could we have tests of full event (`BeaconEvent`) serialization and deserialization too?

torrybr replied:
something like this? Not a complete example, just making sure this is what you are expecting generally

```
#[test]
fn message_event_deserialization() {
    let json_data = json!({
        "content": get_beacon_event_content_json(),
        "event_id": "$beacon_event_id:example.com",
        "origin_server_ts": 1_636_829_458,
        "room_id": "!roomid:example.com",
        "type": "org.matrix.msc3672.beacon",
        "sender": "@example:example.com"
    });

    let event = from_json_value::<AnyMessageLikeEvent>(json_data).unwrap();

    assert_matches!(event, AnyMessageLikeEvent::Beacon(MessageLikeEvent::Original(message_event)));
    assert_eq!(
        message_event.content.relates_to.event_id,
        owned_event_id!("$beacon_info_event_id:example.com")
    );
    assert_eq!(message_event.content.location.uri, "geo:51.5008,0.1247;u=35");
    assert_eq!(message_event.content.ts, MilliSecondsSinceUnixEpoch(uint!(1_636_829_458)));
    assert_matches!(message_event.content.relates_to, Reference { event_id, .. });

    assert_eq!(message_event.sender, "@example:example.com");
    assert_eq!(message_event.room_id, "!roomid:example.com");
    assert_eq!(message_event.event_id, owned_event_id!("$beacon_event_id:example.com"));
    assert_eq!(message_event.origin_server_ts, MilliSecondsSinceUnixEpoch(uint!(1_636_829_458)));
    assert!(message_event.unsigned.is_empty());
}

```

zecakeh replied:
Yes, that would be perfect.

torrybr replied:
Could you point me to a good example for full serialization? Deserialize is complete

zecakeh replied:
Oh sorry, I forgot for a minute that it's not possible to serialize whole events. We'll be fine with just deserialization.

zecakeh reviewed (commented):
Thanks, just a few nits remaining.

torrybr committed:
test(beacon): full message and state deserialization

zecakeh commented:
This looks good, we just need to merge #1839 and rebase for CI to pass. You can squash all the commits, one is enough.

zecakeh commented:
Alright, you just need to rebase now

zecakeh commented:
Thanks a lot!
